ruby/sdk-phase-2.md

```diff
@@ -0,0 +1,605 @@
+    Timeout ::Timeout
+    Mutex = ::Mutex
+    ConditionVariable = ::ConditionVariable
+    Queue = ::Queue
+    SizedQueue = ::SizedQueue
+
+    class ContinueAsNewError < Error
+      def initialize(*args, workflow: nil, **left_off_for_brevity)
+    end
+
+    class Unsafe
+      def replaying?
+      # See TracePoint later on why this is present.
+      def illegal_call_tracing_disabled(&)
+    end
+
+    # Other nested classes omitted for brevity
+  end
+end
+```
+
+Notes:
+
+* Notice lack of `timeout` on `wait_condition`, we will support Ruby `Timeout` via the fiber scheduler.
+* Notice lack of `sleep`, we will support Ruby `sleep` via the fiber scheduler.
+* A `cancellation` can be provided to every async thing, but by default it's the workflow-level one.
+* A workflow builds a `Workflow::Definition::Info` object which contains metadata about the structure including a set of
+  `Workflow::Definition::Signal`s, `Workflow::Definition::Query`s, and `Workflow::Definition::Update`s built from the
+  handlers (but of course handlers can be added/removed at runtime via the `_handlers` hashes).
+* The `workflow_` methods above can only be called during class definition time not workflow runtime.
+* The runtime class methods above can only be called during runtime.
+* `respond_to_missing?` + `inherited` will be leveraged such that all definition-time defined handlers will return their
+  definition if they are called as class methods. So `FooWorkflow.some_signal` will return a
+  `Workflow::Definition::Signal` if an instance method of `some_signal` exists marked as a `workflow_signal`.
+  * 💭 Why? It is helpful for callers/client-side to be able to reference something more explicit than a symbol, and it
+    also helps with making sure if the name is changed we use that name instead of the method name. Symbols can always
+    be used though.
+  * ❓ Is this acceptable or too much magic?
+  * ❓ Is there a better way to reference an interaction outside of the instance of the class that works well with Ruby
+    generic typing frameworks, e.g. Sorbet?
+    * We are trying to avoid the proxy model that Java takes because we have found it is too confusing to users to hide
+      that method invocation is actually, say, activity invocation.
+
+## Dynamic Workflows, Signals, Queries, and Updates
+
+* Support dynamic pieces similar to how we have done in Python and .NET.
+* Create a new `Temporalio::Converters::RawValue` object that is a pass-through object for raw values (still subject to
+  payload codecs), and a `payload_converter` is available in workflow/activity context to convert it.
+  * Handler methods will use `RawValue` for args if told to, similar with workflow `execute`/`initialize` if
+    `workflow_raw_args` is set.
+  * `RawValue` can be returned too.
+* To mark a workflow as dynamic, `workflow_dynamic` class method is called. Only one dynamic workflow can be present for
```

Comment:
I think this is a layering violation.  The workflow itself doesn't need to know whether it's dynamic; instead, it should be registered as dynamic on the worker, which also makes it easier to select which is the dynamic workflow for the worker.  Without that, if the user has two workflows registered as `dynamic_workflow` in the same codebase, it would barf even if only one is registered in the worker.
See https://github.com/coinbase/temporal-ruby/blob/f41efb7bb0c9e755a07382b94e00a4f31bcd33c1/lib/temporal/worker.rb#L97 for what we did in the coinbase SDK.

Reply:
> The workflow itself doesn't need to know whether it's dynamic

This is not true in any SDK nor should it be true here. Dynamic workflows/activities should be allowed to be in the same list at registration time with the rest of the workflows. We want authors to clearly mark and be aware that a workflow, activity, signal, query, or update is dynamic. While technically dynamic workflow and activity's signatures are no different, dynamic signal, query, and update signatures _are_ different and must opt in. So all dynamics in all forms are clearly author opt-in.

> Without that, if the user has two workflows registered as dynamic_workflow in the same codebase, it would barf even if only one is registered in the worker.

No, this is only checked at registration time, same as in every SDK with dynamics. You can have as many as you want implemented, you can only have one dynamic _per worker_. There are no globals here.